ding/policy/wqmix.py

```diff
@@ -0,0 +1,455 @@
+        data['done'] = data['done'].float()
+        return data
+
+    def _forward_learn(self, data: dict) -> Dict[str, Any]:
+        r"""
+        Overview:
+            Forward and backward function of learn mode.
+        Arguments:
+            - data (:obj:`Dict[str, Any]`): Dict type data, a batch of data for training, values are torch.Tensor or \
+                np.ndarray or dict/list combinations.
+        Returns:
+            - info_dict (:obj:`Dict[str, Any]`): Dict type data, a info dict indicated training result, which will be \
+                recorded in text log and tensorboard, values are python scalar or a list of scalars.
+        ArgumentsKeys:
+            - necessary: ``obs``, ``next_obs``, ``action``, ``reward``, ``weight``, ``prev_state``, ``done``
+        ReturnsKeys:
+            - necessary: ``cur_lr``, ``total_loss``
+                - cur_lr (:obj:`float`): Current learning rate
+                - total_loss (:obj:`float`): The calculated loss
+        """
+        data = self._data_preprocess_learn(data)
+        # ====================
+        #  forward
+        # ====================
+        self._learn_model.train()
+        # for hidden_state plugin, we need to reset the main model and target model
+        self._learn_model.reset(state=data['prev_state'][0])
+
+        inputs = {'obs': data['obs'], 'action': data['action']}
+        total_q = self._learn_model.forward(inputs, single_step=False, Q_star=False)['total_q']
+        total_q_star = self._learn_model.forward(inputs, single_step=False, Q_star=True)['total_q']
+
+        next_inputs = {'obs': data['next_obs']}
+        logit_detach = self._learn_model.forward(next_inputs, single_step=False, Q_star=False)['logit'].clone().detach()
+        next_inputs = {'obs': data['next_obs'], 'action': logit_detach.argmax(dim=-1)}
+
+        with torch.no_grad():
+            target_total_q = self._learn_model.forward(next_inputs, single_step=False, Q_star=True)['total_q']
+
+        with torch.no_grad():
+            if data['done'] is not None:
+                target_v = self._gamma * (1 - data['done']) * target_total_q + data['reward']
+            else:
+                target_v = self._gamma * target_total_q + data['reward']
+
+        td_error = (total_q - target_v).clone().detach()
+        _data = v_1step_td_data(total_q, target_total_q, data['reward'], data['done'], data['weight'])
+        _, td_error_per_sample = v_1step_td_error(_data, self._gamma)
+
+        data_star = v_1step_td_data(total_q_star, target_total_q, data['reward'], data['done'], data['weight'])
+        loss_star, td_error_per_sample_star_ = v_1step_td_error(data_star, self._gamma)
+
+        # our implemention is based on the https://github.com/oxwhirl/wqmix
+        # Weighting
+        alpha_to_use = self._cfg.learn.alpha
+        if self._cfg.learn.wqmix_ow:  # Optimistically-Weighted
+            ws = torch.ones_like(td_error) * alpha_to_use
```

Comment:
`torch.full_like` can be better, and it doesn't need additional multiply 1 in `torch.ones_like(td_error) * 1`